src/ocean/util/container/btree/BTree.d

```diff
@@ -0,0 +1,695 @@
+        auto res = deleteFromNode(this.root,
+                parent, element, rebalance_parent);
+
+        // can't rebalance the root node here, as they should all be
+        // rebalanced internally by deleteFromNode
+        assert(rebalance_parent == false);
+
+        debug (BTreeSanity) check_invariants(*this);
+
+        if (res)
+            this.content_version++;
+
+        return res;
+    }
+
+    /******************************************************************************
+
+        Finds the highest value in the tree.
+
+        Returns:
+            pointer to largest element in the tree, or null if the tree is empty.
+
+    ******************************************************************************/
+
+    public Const!(ElementType)* findMaximum ()
+    {
+        size_t index;
+
+        if (this.root is null)
+        {
+            return null;
+        }
+
+        return &(this.impl.findMaximum(this.root, index).elements[index]);
+    }
+
+    /******************************************************************************
+
+        Searches the tree for the element and returns the pointer to it, or null
+        if not found.
+
+        Params:
+            tree = tree to search in
+            element = element to find in a tree.
+
+        Returns:
+            pointer to the element, or null if not found.
+
+    *******************************************************************************/
+
+    public Const!(ElementType)* search (ref ElementType element)
+    {
+        size_t index;
+        auto node = this.impl.search(element, index);
+
+        return node is null? null : &node.elements[index];
+    }
+
+    /**************************************************************************
+
+        Node of the tree. Contains degree * 2 - 1 elements and degree * 2
+        subtrees.
+
+    **************************************************************************/
+
+    package struct BTreeNode
+    {
+        /// Number of the elements currently in the node
+        int number_of_elements;
+
+        /// Indicator if the given node is a leaf
+        bool is_leaf;
+
+        /// Array of the elements
+        package TreeElementType[tree_degree * 2 - 1] elements;
+
+        /// Array of the pointers to the subtrees
+        package BTreeNode*[tree_degree * 2] child_nodes;
+
+        /// Convenience alias of the element type
+        package alias TreeElementType ElementType;
+
+        /// Convenience alias of the tree degree
+        private const degree = tree_degree;
+    }
+
+    /// Convenience alias of the element type
+    package alias TreeElementType ElementType;
+
+    /// degree of a tree
+    package const degree = tree_degree;
+
+    /// Root directory
+    package BTreeNode* root;
+
+    /// "version" of the tree's state. Will be incremented on any removal/adding
+    /// the element. Used to check if the ranges are valid while doing the iteration
+    package int content_version;
```

Comment:
Is a signed integer large enough? If it's incremented on each modification, it wouldn't seem crazy that it could wrap under certain usage conditions.